### ISSUE TITLE
Pause/Resume when a hook is reached

### ISSUE THREAD

**Issue by cryptax** [sensepost/objection#544]
It would be convenient to **pause** when a given hook is reached, so that we can for instance retrieve file from the file system at this exact moment.
For example, if I wish to unpack a sample, I can add a hook onto `DexClassLoader` constructor. When the constructor is reached, if it pauses, I can go and inspect the device, download the DEX. Then, I can resume execution.
Currently, _pause/resume is not possible_, thus, in many cases, I see that `DexClassLoader` is reached, would love to grab the file, but by the time I do, `File.delete()` has been called and I am _too late_.

For example this is what I have:
```
objection --gadget com.qfoqzju.yrcapztuywyu explore --startup-command 'android hooking watch class_method dalvik.system.DexClassLoader.$init --dump-args --dump-return'
(agent) Attempting to watch class dalvik.system.DexClassLoader and method $init.
(agent) Hooking dalvik.system.DexClassLoader.$init(java.lang.String, java.lang.String, java.lang.String, java.lang.ClassLoader)
(agent) Registering job 484905. Type: watch-method for: dalvik.system.DexClassLoader.$init
...
(agent) [025432] Called dalvik.system.DexClassLoader.DexClassLoader(java.lang.String, java.lang.String, java.lang.String, java.lang.ClassLoader)
(agent) [025432] Arguments dalvik.system.DexClassLoader.DexClassLoader(:/data/user/0/com.qfoqzju.yrcapztuywyu/app_files/zczvob.jar, /data/user/0/com.qfoqzju.yrcapztuywyu/app_files, (none), dalvik.system.PathClassLoader[DexPathList[[zip file "/system/framework/org.apache.http.legacy.boot.jar", zip file "/data/app/com.qfoqzju.yrcapztuywyu-Kepiuxeda7PSWVYTzj0XYg==/base.apk"],nativeLibraryDirectories=[/data/app/com.qfoqzju.yrcapztuywyu-Kepiuxeda7PSWVYTzj0XYg==/lib/x86, /system/lib]]])
```

**Describe the solution you'd like**

`objection --gadget com.qfoqzju.yrcapztuywyu explore --startup-command 'android hooking pause class_method dalvik.system.DexClassLoader.$init --dump-args --dump-return'` would be lovely.

Note the keyword `pause`.
Then, when it is paused, I could issue a `resume` to continue when I am ready.

**Describe alternatives you've considered**

I considered hooking `File.delete` and modifying it's return value, but that doesn't work, because there is no meant return value for that. I'd need to make `File.delete` do nothing.

I also considered writing a Frida hook. But the point of easy hooking with objection is not to have to do that :)

**Comment by CDuPlooy** [comment]
I don't think it'd be trivial to add something like that. If instead what your after is a way to pull dex files loaded by `dalvik.system.DexClassLoader` and store them on your testing machine, that could be do-able. 

Maybe something like `android hooking dexloader --dump-dir /tmp` ?

**Comment by leonjza** [comment]
Thinking about this problem today, I was getting stuck on how we would manage to pause everything when an application is running. I don't know the answer that.

However, (ignoring the potentially catastrophic side effects this might have depending on what you are hooking) we could block in a function's `onEnter()` (and probably `onLeave()` too) with something like this:

```javascript
var op = recv('input', function(value) {
    // do work with value.payload
});
op.wait();
```

Maybe this warrants a new flag in the hooking commands to patch in such a blocking `revc()`, even if just to allow for a user to press enter to something to continue?

**Comment by leonjza** [comment]
> My other concern is that it's misleading to say the application has been paused when in reality only that one specific thread is blocking.

Yeah we can just call it a thread block or something more appropriate, with a warning saying the app is _probably_ going to crash. If it does not, buy a lotto ticket! :P

**Comment by CDuPlooy** [comment]
http://www.giovanni-rocca.com/how-i-turn-frick-into-a-real-frida-based-debugger/

I looked into breakpoints a while ago and came across this excellent blog post. There a way to handle process level exceptions (I've got an example somewhere else) that may also be useful. TLDR of what I'm thinking:

* Create a breakpoint using the above
* When it's triggered, suspend the process tree of the app
* Return some context to the objection repl
* Once resume is picked, resume all threads

I think the hardest parts of this would be around ensuring this works on as many android devices as possible. I recall having some trouble with the whole suspending a thread thing.

**Comment by cryptax** [comment]
> I don't think it'd be trivial to add something like that. If instead what your after is a way to pull dex files loaded by `dalvik.system.DexClassLoader` and store them on your testing machine, that could be do-able.
> 
> Maybe something like `android hooking dexloader enable --dump-dir /tmp` ?

Yes, that would work for me, but it's less generic than having a way to issue kind of a breakpoint.
+ note DEX files can also be loaded with PathClassLoader, or InMemoryDexClassLoader.

**Comment by CDuPlooy** [comment]
wow, this year has gone by quickly. 

I've literally spent a hot minute thinking about this, so @leonjza, in line with what you were proposing, how about editing the watch command to add another option `--break` which will set a "BrEaKpOiNt" haha

We could then add something similar to the jobs interface, where for each unique function call, an entry is created and you have to explicitly resume it. This could work using the messaging API combination of send and recv. Something like 

`ios hooking watch NSData --dump-args --break`

Once the method is invoked we can output something like:

`$CLASS & $METHOD has been triggered, blocking until ....` (I'll think of some better wording)

Then users can do `breakpoints list` coupled with `breakpoints continue $BP`

**Comment by IPMegladon** [comment]
Possibly worth looking into this again with https://frida.re/news/2024/09/06/frida-16-5-0-released/ coming out, though not sure how reliable the behaviour on android/ios is going to be when blocking a thread.
